### PR TITLE
Dacheson/ok 37/model validation enforcer

### PR DESCRIPTION
**Pull request by drew-acheson**

### CODE REVIEW

File: src/Enforcers/ModelValidationEnforcer.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+     * @param string $className - The name of the class for which metadata should be
+     * returned.
+     * @return array See ValidationEnforcer@getValidationMetadata.
+     */
+    public function getValidationMetadata(string $className): array
+    {
+        $metadata = parent::getValidationMetadata($className);
+
+        $ref   = new ReflectionClass($className);
+        $props = $ref->getProperties();
```
Comment (bbotto-pdga) [inline src/Enforcers/ModelValidationEnforcer.php:29]: Can't this come from the parent's metadata?  Each property (key) in `$metadata` now has a `reflectionProperty`.  That would avoid creating a second `ReflectionClass` and array of `ReflectionProperty`.

Likewise, the `foreach` loop below could iterate over the `$metadata` getting both the `$propName` and the `$meta` for that property (`validators` and such).

```
foreach($metadata as $propName => $propMeta)
```

File: src/Enforcers/ModelValidationEnforcerTest.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+use PHPUnit\Framework\TestCase;
+
+use PDGA\DataObjects\Attributes\Column;
+use PDGA\DataObjects\Enforcers\ModelValidationEnforcer;
+use PDGA\DataObjects\Validators\EmailValidator;
+use PDGA\DataObjects\Validators\MaxLengthValidator;
+
+class NamedPerson
+{
+    #[Column(
```
Comment (pcrist) [inline src/Enforcers/ModelValidationEnforcerTest.php:12]: Can we get some whitespace separation between properties and attributes in this class?

File: src/Enforcers/ModelValidationEnforcerTest.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        $this->enforcer = new ModelValidationEnforcer();
+    }
+
+    public function testMetadataReturnsCorrectNumberOfValidators(): void
+    {
+        $result = $this->enforcer->getValidationMetadata(NamedPerson::class);
+
+        //id validators include int and not null
+        $this->assertEquals(count($result['id']['validators']), 2);
```
Comment (pcrist) [inline src/Enforcers/ModelValidationEnforcerTest.php:41]: How does `$id` have validators if it doesn't have a Column attribute? Or am I missing something?

Comment (drew-acheson) [reply src/Enforcers/ModelValidationEnforcerTest.php:41]: ModelValidationEnforcer extends ValidationEnforcer.  The ModelValidationEnforcer metadata function calls the parent function first.

File: src/Enforcers/ModelValidationEnforcerTest.php (hunk 2 above)
Comment (drew-acheson) [reply src/Enforcers/ModelValidationEnforcerTest.php:12]: How would you like the spacing to be set?  A space between a property and the beginning of the next propertie's attributes? ie

`
    public int $id;

    #[Column(
        name: 'name',
        maxLength: 20
    )]
    #[MaxLengthValidator(15)]
    public string $name;
`

File: src/Enforcers/ModelValidationEnforcerTest.php (hunk 3 above)
Comment (pcrist) [reply src/Enforcers/ModelValidationEnforcerTest.php:41]: I see. All properties are validated based on their type at least, regardless of whether they have a `Column` attribute.

**Review by bbotto-pdga — APPROVED** [review]
Nice.

**Review by pcrist — APPROVED** [review]

**Pull request merged**